### ISSUE TITLE
Components modifying other components

### ISSUE THREAD

**Issue by Programmerino** [vugu/vugu#109]
**Describe the bug**
Using the same component twice with an intermediate component causes the struct data to be changed only in the first component.

**Software Versions**
Vugu version: master (v0.1.1-0.20200315225605-e910962296fa)
Go version: 1.13.8
Browser and version: Firefox (74.0) & Chromium (80.0.3987.87 (Developer Build))

**To Reproduce**
Below is the most minimal way to reproduce this I could find:
### `root.vugu`
```html
<html>
<body>
<div>
	<main:Helper></main:Helper>
	<br>
	<br>
	<main:Helper></main:Helper>
</div>
</body>
</html>
```
### `helper.vugu`
```html
<div class="helper">
	<main:Final></main:Final>
</div>
<script type="application/x-go">
	type Helper struct {
		UselessVariable string // See explanation in additional notes
	}
</script>
```
### `final.vugu`
```html
<div class="final">
	<div @click='c.Clicked=true'>Click me</div>
	<div vg-if='c.Clicked'>
		<p>You clicked me!</p>
	</div>
	<div vg-if='!c.Clicked'>
		<p>I haven't been clicked!</p>
	</div>
</div>

<script type="application/x-go">
	type Final struct {
		Clicked bool
	}
</script>
```
**Expected behavior**
I would expect that clicking the "Click me" would change the `p` underneath to `You clicked me!` within the component, however, this is what actually happens:
* Clicking in the second component only changes the first
* Clicking the first without first having clicked the second does nothing
* Clicking the first after clicking the second reverts it back to "I haven't been clicked!" (theoretically, a click should be impossible to undo with how the application works)

**Additional Notes**
* If you directly reference final from root, the bug does not occur
* The intermediate component "helper" must have a struct defined with at least one field, otherwise clicking does nothing to either components (is this a bug?)
* My devserver is stock from the vugu guide, no modifications

**Comment by jyellick** [comment]
+1 Just ran into this myself.

TLDR; You can work around this using the latest master by specifying a `vg-key` attribute on your element.  Simply modify your code to look like:

**root.vugu**
```
<html>
<body>
<div>
  <main:Helper :UselessVariable='"key1"'></main:Helper>
  <br>
  <br>
  <main:Helper :UselessVariable='"key2"'></main:Helper>
</div>
</body>
</html>
```
**helper.vugu**
```
<div class="helper">
  <main:Final vg-key='c.UselessVariable'></main:Final>
</div>
<script type="application/x-go">
  type Helper struct {
    UselessVariable string // Actually not so useless :)
  }
</script>
```

----

The underlying problem seems to be that the component ID is only generated once per type that the component is embedded in.

https://github.com/vugu/vugu/blob/7d3a6c0162c2fca568a4bf8d6c7163166f7b45fe/gen/parser-go.go#L908

And when we go to look it up in the cache:

https://github.com/vugu/vugu/blob/7d3a6c0162c2fca568a4bf8d6c7163166f7b45fe/gen/parser-go.go#L911

Since this is a static string, that's shared by all instances of the component, when we get to this code:

https://github.com/vugu/vugu/blob/7d3a6c0162c2fca568a4bf8d6c7163166f7b45fe/build-env.go#L124-L131

The first time through we get an existing instance, and the second time through, there's nothing in the cache, so we get a new instance (losing all of our internal state).

Similarly, when we use the component:

https://github.com/vugu/vugu/blob/7d3a6c0162c2fca568a4bf8d6c7163166f7b45fe/gen/parser-go.go#L917

in the builder code we end up overriding the first instance with the second: 

https://github.com/vugu/vugu/blob/7d3a6c0162c2fca568a4bf8d6c7163166f7b45fe/build-env.go#L133-L139

I noticed that there's a vgKeyExpr call here:

https://github.com/vugu/vugu/blob/7d3a6c0162c2fca568a4bf8d6c7163166f7b45fe/gen/parser-go.go#L904

So my inference would be that perhaps, the user is expected to supply a disambiguating attribute in cases like these.  Although my React experience is novice at best, I seem to recall the tooling spitting out warnings requiring that a unique ID by attached to certain elements so that the framework could differentiate them.  Maybe @bradleypeabody can weigh in, as to whether things are working as designed here, and it's expected that users supply a `vg-key`, or, if this is actually a bug.

**Comment by bradleypeabody** [comment]
Yeah this is definitely a bug.  What is supposed to happen is that the position of the reference in the document plus its type are hashed together to create a unique ID.  Looks like this logic got lost in one of the refactors and it's only using the type.

`vg-key` is meant for cases where there is a list of component references in sequence so it can re-use components that change sequence.

I haven't totally nailed down a clear statement of exactly how the component re-use logic should be.  That would be a good start to fully addressing this issue.

**Comment by bradleypeabody** [comment]
I looked into this more closely and @jyellick I understand what you mean now by:
> the component ID is only generated once per type that the component is embedded in.

The core problem is that right now component reference ID numbers are generated statically at compile time and are meant to give a unique ID to each instance of the component.  As it turns out this doesn't actually work because it can only express the parent-child relationship that components have directly in one file.  These IDs do not, for example, reflect different instances underneath different parents.  This is also why examples that use one level of component nesting work correctly but this one uses an additional level and is broken.

I think the solution has to be to get an identifier from the BuildEnv that is unique to each branch of the build tree and hash that into the existing static values.  So the uniqueness properties would be what they are now plus the overall "position" in the tree where it was rendered.  (Calculating "position" has specific logic I've used elsewhere and helps avoid unrelated branches from unnecessarily changing hash values, at least somewhat - I'll drop that in here as well.)  This should have the desired effect.  I'll post a followup as soon as I have that in place.

**Comment by bradleypeabody** [comment]
Yup, that did the trick.  This is fixed now on master, will cut a new release shortly.

(Basically those component references now have `^vgin.CurrentPositionHash()` added to the hash, which provide the logic above.)
